Adds Playwright documentation (#21402)
cloudflare/cloudflare-docs: src/content/docs/browser-rendering/platform/playwright.mdx

---
pcx_content_type: concept
title: Playwright
description: Learn how to use Playwright with Cloudflare Workers for browser automation. Access Playwright API, manage sessions, and optimize browser rendering.
sidebar:
  order: 5
  badge: Beta
---

import { Render, WranglerConfig, TabItem, Tabs } from "~/components";

[Playwright](https://playwright.dev/) is an open-source package developed by Microsoft that can do browser automation tasks; it's commonly used to write frontend tests, create screenshots, or crawl pages.

The Workers team forked a [version of Playwright](https://github.com/cloudflare/playwright) that was modified to be compatible with [Cloudflare Workers](https://developers.cloudflare.com/workers/) and [Browser Rendering](https://developers.cloudflare.com/browser-rendering/).

> Our version is open sourced and can be found in [Cloudflare's fork of Playwright](https://github.com/cloudflare/playwright). The npm can be installed from [npmjs](https://www.npmjs.com/) as [@cloudflare/playwright](https://www.npmjs.com/package/@cloudflare/playwright):

Reviewer: ```suggestion
Our version is open sourced and can be found in [Cloudflare's fork of Playwright](https://github.com/cloudflare/playwright). The npm package can be installed from [npmjs](https://www.npmjs.com/) as [@cloudflare/playwright](https://www.npmjs.com/package/@cloudflare/playwright):
```

    await page.goto('https://demo.playwright.dev/todomvc');

    const TODO_ITEMS = todos.length > 0 ? todos : [
      'buy some cheese',
      'feed the cat',
      'book a doctors appointment'
    ];

    const newTodo = page.getByPlaceholder('What needs to be done?');
    for (const item of TODO_ITEMS) {
      await newTodo.fill(item);
      await newTodo.press('Enter');
    }

    await expect(page.getByTestId('todo-title')).toHaveCount(TODO_ITEMS.length);

    await Promise.all(TODO_ITEMS.map(
      (value, index) => expect(page.getByTestId('todo-title').nth(index)).toHaveText(value)
    ));

    if (trace) {
      await page.context().tracing.stop({ path: 'trace.zip' });
      await browser.close();
      const file = await fs.promises.readFile('trace.zip');

      return new Response(file, {
        status: 200,
        headers: {
          'Content-Type': 'application/zip',
        },
      });
    } else {
      const img = await page.screenshot();
      await browser.close();

      return new Response(img, {
        headers: {
          'Content-Type': 'image/png',
        },
      });
    }
  },
};
```

### Assertions

> One of the most common use cases for using Playwright is software testing. Playwright includes test assertion features in its APIs; see [this page](https://playwright.dev/docs/test-assertions) for detailed documentation. Here's an example of a Worker doing `expect()` test assertions of the [todomvc](https://demo.playwright.dev/todomvc) demo page:

Reviewer: ```suggestion
One of the most common use cases for using Playwright is software testing. Playwright includes test assertion features in its APIs; refer to [Assertions](https://playwright.dev/docs/test-assertions) in the Playwright documentation for details. Here's an example of a Worker doing `expect()` test assertions of the [todomvc](https://demo.playwright.dev/todomvc) demo page:
```
